map3d.py

```diff
@@ -0,0 +1,133 @@
+
+
+
+def map3d_surface(xdata,ydata,zdata,scalardata,mode,vmin,vmax,data_cmap='blue-red',data_alpha=1,topo=None,topo_limits=None,zscale=500.,topo_vmin=None,topo_vmax=None,topo_cmap='bone',topo_cmap_reverse=False,land_constant=False,land_color=(0.7,0.7,0.7),set_view=None):
+    """
+    xdata = longitude values for data array
+    ydata = latitude values for data array
+    zdata = depth values for data array
+    vmin = colorbar minimum for data
+    vmax = colorbar maximum for data
+    data_cmap = colormap for data surface, default is blue-red
+    data_alpha = opacity for data surface from 0 to 1, default is 1
+    topo = optional; input topography file, default is etopo 30 
+    topo_limits = optional; longitude and latitude limits for 3d topography plot [lon min, lon max, lat min, lat max], longitudes range -180 to 180, latitude -90 to 90, default is entire globe
+    zscale = optional; change vertical scaling for plotting, default is 500
+    topo_cmap = optional; default is bone 
+    topo_cmap_reverse = optional; reverse topography colormap, default is false
+    set_view = optional; set the mayavi camera angle with input [azimuth, elevation, distance, focal point], default is 
+    """
+    
+    #import dependent modules
+    import sys
+    import numpy as np
+    from mayavi import mlab
+
+    #load topo data
+    data = np.load('etopo1_30min.npz')
+    xraw = data['x']
+    yraw = data['y']
+    zraw = np.swapaxes(data['z'][:,:],0,1)
+    zraw[zraw>0]=0.
+    phi = (yraw[:]*np.pi*2)/360.+np.pi/2.
+    theta = (xraw[:]*np.pi*2)/360.
+    c = zraw
+    theta=np.append(theta,theta[0])
+    c = np.concatenate((c,np.expand_dims(c[0,:],axis=0)),axis=0)
+
+    if topo_limits is None:
+        #set default limits (phi=latitude, theta=longitude)
+        phi_1 = -90
+        phi_2 = 90
+        theta_1 = -180
+        theta_2 = 180
+    
+    else:
+        phi_1 = topo_limits[2]
+        phi_2 = topo_limits[3]
+        theta_1 = topo_limits[0]
+        theta_2 = topo_limits[1]
+    
+    #find indices of topo limits
```

Comment:
Each time you comment a block like this, it might be a candidate for a helper function.
That way instead of reading one large function by reading one block comment to another, you can read it through at a higher level of abstraction, like:
```
topo_data = load_topo_data(topo_file_name)          # ~ lines 26-36
topo_data = crop_dataframe(topo_data, topo_limits)  # ~ lines 51-64

if mode == 'sphere':
    transform = transform_sphere
elif:
    ...

topo_data = transform(topo_data)

x, y, z = topo_data
m = mlab.mesh(x, y ,z , colormap=topo_cmap, vmin=tvmin / zscale, vmax=tvmax / zscale)
...
```